src/main/java/org/simlar/simlarserver/services/accountservice/AccountService.java

```diff
@@ -278,19 +299,29 @@ SimlarId call(final String telephoneNumber, final String password, final Instant
         return simlarId;
     }
 
-    public void confirmAccount(final String simlarIdString, @SuppressWarnings("TypeMayBeWeakened") final String registrationCode) {
+    public void confirmAccount(final String simlarIdString, final String registrationCode) {
         final SimlarId simlarId = SimlarId.create(simlarIdString);
         if (simlarId == null) {
             throw new XmlErrorNoSimlarIdException("confirm account request with simlarId: " + simlarIdString);
         }
 
+        confirmAccountRequest(AccountRequestType.CREATE, simlarId, registrationCode);
+    }
+
+    public void confirmAccountDeletion(final String telephoneNumber, final String registrationCode) {
+        confirmAccountRequest(AccountRequestType.DELETE, createValidatedSimlarId(telephoneNumber), registrationCode);
+    }
+
+    @SuppressFBWarnings("NP_NULL_ON_SOME_PATH_EXCEPTION") // false positive
+    @SuppressWarnings("OverlyComplexMethod")
+    private void confirmAccountRequest(final AccountRequestType type, final SimlarId simlarId, @SuppressWarnings("TypeMayBeWeakened") final String registrationCode) {
         if (!checkRegistrationCodeFormat(registrationCode)) {
             throw new XmlErrorNoRegistrationCodeException("confirm account request with simlarId: " + simlarId + " and registrationCode: " + registrationCode);
         }
 
         final AccountCreationRequestCount creationRequest = transactionTemplate.execute(status -> {
             final AccountCreationRequestCount dbEntry = accountCreationRepository.findBySimlarIdForUpdate(simlarId.get());
```

Comment:
Maybe something like so that the `|| dbEntry.getType() != type` could be removed
```suggestion
            final AccountCreationRequestCount dbEntry = accountCreationRepository.findBySimlarIdAndTypeForUpdate(simlarId.get(), type);
```